### DISCUSSION TITLE
Constructor inlining

### DISCUSSION THREAD

**Discussion by scf37** [scala/scala3#17722] [Feature Requests]
Dotty has method inlining but it is incomplete without constructor inlining.

Motivation
========
```scala
trait JsonReader[A] {
  def read(it: JsonIterator): A
}

class SeqJsonReader[A](inner: JsonIterator => A) {
  inline def read(it: JsonIterator): A = ..... // won't inline as expected because constructor parameters are not inlined
}

inline given [A]: SeqJsonReader[A] = new SeqJsonReader(it => summon[JsonReader[A]].read(it)) // summoned read() contents will be inlined

```
Problem
======

```scala
  class Foo (a: String) {
    inline def foo(): Unit = println(a)
  }

  def print() = new Foo("hello").foo()
```
Compiles to:
```java
public void print() {
    Foo Foo_this = new Foo("hello");
    Predef$.MODULE$.println((Object)Foo_this.inline$a());
}
```
which is wasted allocation.

Therefore when inlining method referencing constructor parameters only, there is no need to create object instance. 

Possible solutions
==============
1. Inline classes - kind of explored here https://odersky.github.io/sips//pending/inline-classes.html
2. Perform this optimization automatically

**Comment by TheElectronWill** [comment]
This is related to the idea of an "inliner-optimizer" (quoted in [this thread about for loops](https://github.com/lampepfl/dotty/issues/8880)) that would perform scalar replacement (replacing the fields by local variables and inlining the constructor). I'm actually playing with the idea by implementing a new compiler phase in dotty.
